lib/elixir/lib/module.ex

```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
+             "is always ignored for private functions"}
+  end
+  defp do_check_impl(_, kind, _, _, _) when kind in [:type, :typep] do
+    {:error, "@impl is always ignored for type annotations"}
+  end
+  defp do_check_impl(_, _, _, value, impl_required_for_callbacks?)
+       when value in [nil, false] and not impl_required_for_callbacks? do
+    :ok
+  end
+  defp do_check_impl(module, _, function_tuple, value, impl_required_for_callbacks?)
+       when value in [nil, false] and impl_required_for_callbacks? do
+    callbacks_with_behaviour = for behaviour <- get_behaviours(module),
+                                   callback <- behaviour.behaviour_info(:callbacks),
+                                   into: %{},
+                                   do: {callback, behaviour}
+
+    if %{^function_tuple => behaviour} = callbacks_with_behaviour do
+      message = "module attribute @impl was not set for implemented callback " <>
+                "#{format_function(function_tuple)} (callback specified in " <>
+                "#{inspect(behaviour)})"
+      {:error, message}
+    else
+      :ok
+    end
+  end
+  defp do_check_impl(module, _kind, function_tuple, value, _) when value == true do
+    behaviours = get_behaviours(module)
+    callbacks_with_behaviour = for behaviour <- behaviours,
+                                   defines_callbacks?(behaviour),
+                                   callback <- behaviour.behaviour_info(:callbacks),
+                                   into: %{},
+                                   do: {callback, behaviour}
+    cond do
+      :maps.is_key(function_tuple, callbacks_with_behaviour) ->
+        :ok
+      true ->
+        {
+          :error,
+          msg_impl_true_but_function_is_not_callback(function_tuple, callbacks_with_behaviour, behaviours),
+        }
+    end
+  end
+  defp do_check_impl(module, _kind, function_tuple, behaviour, _) when is_atom(behaviour) do
+    cond do
+      not Code.ensure_compiled?(behaviour) ->
+        {:error, "got @impl #{inspect(behaviour)} but #{inspect(behaviour)} " <>
+                 "is not defined"}
+      not defines_callbacks?(behaviour) ->
+        {:error, "got @impl #{inspect(behaviour)} but #{inspect(behaviour)} " <>
+                 "does not define any callbacks"}
+      not behaviour in get_attribute(module, :behaviour) ->
+        {:error, "module attribute @impl was set but this module does not " <>
+                 "implement #{inspect(behaviour)}. Did you forget to specify" <>
+                 " @behaviour #{inspect(behaviour)} ?"}
+      true ->
+        callbacks = behaviour.behaviour_info(:callbacks)
+        if function_tuple in callbacks do
+          :ok
+        else
+          {
+            :error,
+            msg_impl_behaviour_but_function_is_not_callback(function_tuple, callbacks, behaviour),
+          }
+        end
+    end
+  end
+
+  defp to_sentence(options, last_words_connector \\ " or ")
```

Comment:
This `to_sentence` function seems a bit too much for `Module`, but I couldn't think of another way to nicely present the warnings.